### PR TITLE
Run cancel calls on the Zipline dispatcher

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/OutboundCallHandler.kt [hunk 1]
```diff
@@ -42,21 +43,24 @@ internal class OutboundCallHandler(
...
-    val encodedCall = endpoint.json.encodeToStringFast(endpoint.callSerializer, call)
-
-    val callStart = endpoint.eventListener.callStart(call)
-    val encodedResult = endpoint.outboundChannel.call(encodedCall)
-
-    val result = endpoint.json.decodeFromStringFast(function.kotlinResultSerializer, encodedResult)
-    endpoint.eventListener.callEnd(call, result, callStart)
-    return result.getOrThrow()
+    val externalCall = endpoint.callCodec.encodeCall(internalCall, service)
+    val callStart = when (service) {
```
Comment (adrw) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/OutboundCallHandler.kt]: I wonder if a `fun Service.ifNotSuspendCallback(block: Unit) {}` would work better here than the 4 when blocks I've seen so far

File: zipline/src/engineTest/kotlin/app/cash/zipline/EventListenerEndpointTest.kt [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    assertEquals(
+      "fun transform(kotlin.String, app.cash.zipline.testing.EchoService): app.cash.zipline.testing.EchoService",
+      serviceCall.functionName,
+    )
+    assertEquals("hello", serviceCall.args[0])
+    assertNotSame(serviceCall.args[1], serviceArgument) // The service receives a stub.
+    assertTrue(serviceCall.args[1] is EchoService)
+    assertEquals(
+      """
+      |{
```
Comment (adrw) [inline zipline/src/engineTest/kotlin/app/cash/zipline/EventListenerEndpointTest.kt]: Great snapshot & functionality tests

**Review by adrw — APPROVED** [review]

**Review by Norwood6731 — APPROVED** [review]

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/OutboundCallHandler.kt (hunk 1 above)
Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/OutboundCallHandler.kt]: Ooooh I’m generally reluctant to introduce control flow things like that? But it does sound tempting

Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/OutboundCallHandler.kt]: Gonna pass for now!

**Pull request merged**